rfcs/0000-plan-dynanism.md

```diff
@@ -0,0 +1,251 @@
+   where the first path is a derivation, and the second is the output we want to build.
+
+   We should generalize the grammar like so:
+   ```
+   single-installable ::= <single-installable> ! <output-name>
+                       |  <path>
+
+   multi-installable  ::= <single-installable>
+                       |  <single-installable> ! *
+   ```
+
+   Plain paths just mean that path itself is the goal, while `!` indexing indicates one more outputs of the derivation to the left of the `!` is the goal.
+
+   > For example,
+   > ```
+   > nix build /nix/store/…foo.drv
+   > ```
+   > would just obtain `/nix/store/…foo.drv` and not build it, while
+   > ```
+   > nix build /nix/store/…foo.drv!*
+   > ```
+   > would obtain (by building or substituting) all its outputs.
+   > ```
+   > nix build /nix/store/…foo.drv!out!out
+   > ```
+   > would obtain the `out` output of whatever derivation `/nix/store/…foo.drv!out` produces.
+
+   Now that we have `path` vs `path!*`, we also don't need `--derivation` as a disambiguator, and so that should be removed along with all the complexity that goes with it.
+   (`toDerivedPathsWithHints` in the the nix commands should always be pure functions and not consult the store.)
+
+3. Extend the scheduler and derivation dependencies similarly:
+
+  - Derivations can depended on the outputs of derivations that are themselves derivation outputs.
+    The schedule will substitute derivations to simplify dependencies as computed derivations are built, just like how floating content addressed derivations are realized.
+
+  - Missing derivations get their own full fledged goals so they can be built, not just fetched from substituters.
+
+4. Add a new `outputOf` primop:
+
+   `builtins.outputOf drv outputName` produces a placeholder string with the appropriate string context to access the output of that name produced by that derivation.
+   The placeholder string is quite analogous to that used for floating content addressed derivation outputs.
+   This isn't needed today when we statically know the derivation and thus it's outputs to make attributes for them, but it is needed for built derivations because we don't know either the final derivation or its output names in advanced.
+
+## Deferred import from derivation.
+
+Create a new primop `assumeDerivation` that takes a single expression.
+It's semantics are as follows:
+
+ - If the underlying expression evaluates (shallowly) without needing to build derivations (for `import`, `readFile`, etc.), and is a derivation, simply return that.
+   ```
+   e ⇓ e'
+   e' is derivation
+   ------
+   builtins.assumeDerivation e ⇓ e'
+   ```
+
+- If the underlying expression cannot evaluate (shallowly) without building one or more paths, defer evaluation into a derivation-producing-derivation, and take it's output with `builtins.outputOf`:
+  ```
+  e gets stuck on builds
+  defer = derivation {
+    inputDrvs = builds;
+    buildCommand = "nix-instantiate ${reified e} > $out"
```

Comment:
Is `nix-instantiate` supposed to work in that derivation? Is `nix` always available in the build?

Reply:
I am not sure of that? Evaluation is current designed so that like it or not, writing drv files is a side effect.

Reply:
Note that at the time of writing I was not aware of recursive nix being possible in `nixUnstable`. Am I correct in assuming that this hypothetical derivation would use recursive nix to generate the .drv file?

Reply:
Yup. I would want to limit recursive nix so one can just add paths to store (including DRVs), not build things.

Reply:
Okay, I think that should at least be mentioned for clarity. If I understand it correctly, then maybe the following derivation would be clearer:

```nix
derivation {
  buildCommand = "cp $(nix-instantiate ${reifed e}) $out";
}
```

In this case, the inner (recursive) instantiation would add the derivation file to the outer store, but due to content addressability of derivations, its `outPath` would resolve to the same file it produced as a side effect, right?

Reply:
This reification is rather hand-wavy.

The language implementation can't serialize arbitrary closures. Such a feature isn't impossible but has consequences:
 - We'll need a serialization format that is stable, because when used in a non-floating derivation, hashes will be based off it. Changes to the format will not be acceptable or Nix needs to let go of its stability promises.
 - While thunks are fairly small, anything that's been evaluated a bit further in the entire scope will have to be serialized. This will tend to include the attributes of nixpkgs, including a serialization of each evaluated package's attrsets etc.
 - That's assuming serializing thunks shallowly is even acceptable. I don't think it is, because we can't let the serialization depend on the evaluation order. We can't have `nix build .` produce a different derivation for `z` than `nix build .#z`.
 - The closure of an expression may contain references to files that only the Nix _client_ is normally able to access. This applies to local paths. I don't think we want to send the contents of `~` along. It also applies to the general context of the Nix client, like git credentials, Nix plugins, etc. (indeed mostly represented by files, but accessed through abstractions like the builtin fetchers)
 - The closure may contain confidential values, such as NixOps' `deployment.keys.<name>.text` or similar. We don't want those in the store.

Compared to a somewhat naive serialization, another partial solution is to do "event sourcing" in the spirit of [transient (Haskell)](https://hackage.haskell.org/package/transient). To implement that, the evaluator must only be accessible through a facade interface that tracks exactly how an evaluation came to be: which files were provided to it and which external traversals and function calls were made, etc. This method, if it even applies here, seems to solve the thunk dilemma, but doesn't solve the other problems and requires significant changes to the evaluator. The other problems remain unsolved.

Is not serializing the closure an option? We could suspend the evaluation by just retaining the reference to the thunk, assigning it a random id and wait for the daemon to get back to us. This defeats the desired separation between the evaluation and build processes. It does let a derivation like `fetchNodeModules lockFile` access confidential git repos via `builtins.fetchGit`, if the connected client has access to those. That'd be neat.

I hope I've just misunderstood something, because I'd like this to succeed. I just don't see a good way to implement this.
I feel like the old `ret-cont` is conceptually simpler, doesn't create this problem and doesn't create unrealistic expectations.